core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/host/decorator/TelemetryDecorator.java

```diff
@@ -56,22 +97,108 @@ public void decoratePillar(Telemetry telemetry, String clusterName, StackType st
+            if (common != null) {
+                if (StringUtils.isNotEmpty(common.getUser())) {
+                    user = common.getUser();
+                }
+                if (StringUtils.isNotEmpty(common.getGroup())) {
+                    group = common.getGroup();
+                }
+            }
+            fluentConfig.put(FLUENT_USER_PROPERTY, user);
+            fluentConfig.put(FLUENT_GROUP_PROPERTY, group);
+
+            fluentConfig.put(FLUENT_SERVER_LOG_PREFIX_PROPERTY, LOGS_FOLDER_PREFIX_DEFAULT);
+            fluentConfig.put(FLUENT_AGENT_LOG_PREFIX_PROPERTY, LOGS_FOLDER_PREFIX_DEFAULT);
+            fluentConfig.put(FLUENT_SERVICE_LOG_PREFIX_PROPERTY, LOGS_FOLDER_PREFIX_DEFAULT);
+            fluentConfig.put(FLUENT_ENABLED_PROPERTY, true);
+
+            if (LoggingOutputType.S3.equals(logging.getOutputType()) && attributes.getS3Attributes() != null) {
+                S3LoggingAttributes s3Attributes = attributes.getS3Attributes();
+                fluentConfig.put(FLUENT_PROVIDER_PREFIX_PROPERTY, "s3");
+                fluentConfig.put(S3_PARTITION_INTERVAL_PROPERTY, defaultIfNull(s3Attributes.getPartitionIntervalMin(), PARTITION_INTERVAL_MIN_DEFAULT));
+                calculateS3BucketAndLogFolder(clusterName, stackType, s3Attributes, fluentConfig);
+            } else {
+                fluentConfig.put(FLUENT_ENABLED_PROPERTY, false);
             }
-            String defaultLogFolder = Paths.get("cluster-logs", clusterType, clusterName).toString();
-            fluentConfig.put("s3LogFolderName", attributes.getOrDefault("s3LogFolderName", defaultLogFolder));
             servicePillar.put("fluent",
                     new SaltPillarProperties("/fluent/init.sls", singletonMap("fluent", fluentConfig)));
         }
     }
+
+    @VisibleForTesting
+    void calculateS3BucketAndLogFolder(String clusterName, StackType stackType,
+            S3LoggingAttributes attributes, Map<String, Object> fluentConfig) {
+        String bucket = attributes.getBucket();
+        String basePath = attributes.getBasePath();
+        BucketFolderPrefixPair bucketFolderPrefixPair = generateBucketFolderPrefixPair(bucket, basePath);
+
+        String clusterType = CLUSTER_TYPE_DISTROX;
+        if (StackType.DATALAKE.equals(stackType)) {
+            clusterType = CLUSTER_TYPE_SDX;
+        }
+        String s3LogFolderName = Paths.get(CLUSTER_LOG_PREFIX, clusterType, clusterName).toString();
+        if (StringUtils.isNotEmpty(bucketFolderPrefixPair.getFolderPrefix())) {
+            s3LogFolderName = Paths.get(bucketFolderPrefixPair.getFolderPrefix(), clusterType, clusterName).toString();
+        }
+
+        fluentConfig.put(S3_LOG_ARCHIVE_BUCKET_PROPERTY, bucketFolderPrefixPair.getBucket());
+        fluentConfig.put(S3_LOG_FOLDER_NAME_PROPERTY, s3LogFolderName);
+    }
+
+    private BucketFolderPrefixPair generateBucketFolderPrefixPair(String bucket, String basePath) {
```

Comment:
Is this method well tested in the `TelemetryDecoratorTest`? I think there are no tests for weird inputs and miss-formatted bucket urls. My suggestion is to extract this logic to a new class called `BucketFolderProvider` or something like that and test it directly. This logic screams for a separate class.

Reply:
i will create a new jira about validate on request level